### ISSUE TITLE
Dict key assignment narrows the type of this key

### ISSUE THREAD

**Issue by sshishov** [microsoft/pyright#10720]
This probably not a bug but maybe a question.

Imagine we have this structure:
```python
import typing as t


myDict: dict[str, t.Any] = {'key1': 'value'}
myDict['key2'] = 10
t.reveal_type(myDict)
t.reveal_type(myDict['key1'])
t.reveal_type(myDict['key2'])
```
And running pyright giving the following types:
```
information: Type of "myDict" is "dict[str, Any]"
information: Type of "myDict['key1']" is "Any"
information: Type of "myDict['key2']" is "Literal[10]"
```
The questions are:
- why the type of `key2` value is not `Any`?
- if it is expected then how to make the `key2` value have type `Any` without extra variable usage?

**NOTE**: the example is very simplified and the real example contains "complex" types which are inferred and then `pyright` complaining everywhere because of usage. The original dictionary was forcelly assigned type to have values as `Any`...

Best regards, Sergei

**Comment by Glinte** [comment]
Edit: hmm, pyright does say "In general, the type `Any` is not narrowed" ([source](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=narrowing-any)). I wonder if having narrowing here is intended.

My guess is that key2 is Literal[10] because pyright has special narrowing logic for this sort of statically knowable literal assignments. The first thing I think is that if pyright complain about your use of `myDict["key2"]`, maybe pyright is actually right here and other parts of the code are annotated incorrectly, or making incorrect assumptions of the type of `myDict["key2"]`. I personally like that pyright infers exact literal types where possible.

Anyways, if you really want to make pyright not able to do type narrowing here. Try this:

```python
import typing as t

myDict: dict[str, t.Any] = {'key1': 'value'}
myDict.update(key2=10)
# or: myDict.update({"key2": 10})
```

**Comment by erictraut** [comment]
This question relates to the behavior of [type narrowing](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing) on assignment. In general, when a value is assigned to an expression, that expression's type is narrowed based on the type of the assigned value. This narrowing behavior isn't mandated (or really discussed in any way) by the typing spec, so it's left to each type checker to decide whether and how to implement this.

When pyright handles assignments to name expressions (e.g. `x = 1`) or attribute access expressions (e.g. `x.y = 1`), its narrowing logic first determines whether the target symbol has a declared type. If that declared type includes an explicit `Any`, it attempts to retain the `Any` when narrowing. The exact algorithm it uses here is not documented, and I've tweaked it over time. It's full of heuristics and hard-to-defend edge cases — not something I'm particularly proud of, nor something I'd recommend that we put in the typing spec. That said, I would welcome more standardization in this area, so if you have opinions here and would like to propose an addition to the spec, please do so.

When pyright handles assignments to subscripted expressions (e.g. `x["name"] = 1`), there is no symbol involved in the subscript. You might argue that if `x` is a `dict` instance that has a specialized type, a type checker could look at the second type argument and see if it is `Any`, but that involves hard-coding knowledge about `dict` specifically — not something a type checker should do. A more generalized solution is to look at the specialized type of the `__setitem__` method and its parameters, but that's getting pretty complex, especially if `__setitem__` is overloaded. For these reasons, pyright doesn't currently apply any of the special `Any` narrowing logic for subscripted expressions. It simply narrows based on the assigned type. That's why you're seeing this behavior in your code sample.

I think this is the first time I've received a question about this behavior, so I don't think it has been an issue for users. It's likely that many users prefer the current behavior and would see it as a regression if I were to change it. Given the complexity of a potential fix and the other tradeoffs involved, I'd need to see a strong signal from pyright users (or guidance in the typing spec) before making such a change.

**Comment by pinterior** [comment]
type inference by assignment has never caused problems for me, but it seems that mixing subscription assignments and other dict-modification methods may cause some type confusion.

```py
from typing import Any, reveal_type

a: dict[str, Any] = {}
a["foo"] = 12
a["bar"] = 34
reveal_type(a["foo"])
reveal_type(a["bar"])

a.update({"foo": 567})
reveal_type(a["foo"])

a.clear()
reveal_type(a["bar"])
```
```
Type of "a["foo"]" is "Literal[12]"
Type of "a["bar"]" is "Literal[34]"
Type of "a["foo"]" is "Literal[12]"
Type of "a["bar"]" is "Literal[34]"
```

**Comment by sshishov** [comment]
Thank you @Glinte for the working workaround. I will do this in our codebase where we do not want the "narrowed" type for the dict children. The issue happens when we are using the assignment solution to prepare some complex dict using lists and other dicts as children and then use direct `index` or `getitem` (mainly it is used inside the tests, where you prepare request data and then validate the response based on this data).

Thank you @erictraut for yout in-depth and detailed explanation. It seems pretty clear from your explanation and I understand the complexity of type inferring. What I think (at least for dictionary) - infer type of complex dict (where children are lists, dicts, objects, primitives) become useless as we cannot use `dict['key'][0]['name']` later on because the usage `...[0]` will be flagged because of primitives and dicts. The only solution here I can see is to infer not the union of types, but the typeddict with specific types.

@pinterior I guess this will be WorksAsExpected as the type is inferred ONLY during initial assignment. You cannot change the type afterwards. If you want to make it less restrictive, you should do it like:

```python
a: dict[str, Any] = {}
a.update({'foo'})
```
or
```python
a: dict[str, Any] = {}
foo: int = 12
a["foo"] = foo  # in this case the reveal will be `int` I guess
```

Anyhow, working with complex dict types without proper types assignment - is getting quiet difficult, imho. TypedDicts can help, but I do not think that we can "infer" typeddicts 😕

**Comment by erictraut** [comment]
I'm going to mark the issue as closed because I think the original question was answered, and there's nothing actionable at this time.
